src/Shared/HttpSys/RequestProcessing/NativeRequestContext.cs

```diff
@@ -17,24 +18,45 @@ namespace Microsoft.AspNetCore.HttpSys.Internal
     internal unsafe class NativeRequestContext : IDisposable
     {
         private const int AlignmentPadding = 8;
+        private const int DefaultBufferSize = 4096 - AlignmentPadding;
         private IntPtr _originalBufferAddress;
         private HttpApiTypes.HTTP_REQUEST* _nativeRequest;
-        private byte[] _backingBuffer;
+        private IMemoryOwner<byte> _backingBuffer;
+        private MemoryHandle _memoryHandle;
         private int _bufferAlignment;
         private SafeNativeOverlapped _nativeOverlapped;
         private bool _permanentlyPinned;
+        private bool _disposed;
 
         // To be used by HttpSys
-        internal NativeRequestContext(SafeNativeOverlapped nativeOverlapped,
-            int bufferAlignment,
-            HttpApiTypes.HTTP_REQUEST* nativeRequest,
-            byte[] backingBuffer,
-            ulong requestId)
+        internal NativeRequestContext(SafeNativeOverlapped nativeOverlapped, MemoryPool<Byte> memoryPool, uint? bufferSize, ulong requestId)
         {
             _nativeOverlapped = nativeOverlapped;
-            _bufferAlignment = bufferAlignment;
-            _nativeRequest = nativeRequest;
-            _backingBuffer = backingBuffer;
+
+            // TODO:
+            // Apparently the HttpReceiveHttpRequest memory alignment requirements for non - ARM processors
+            // are different than for ARM processors. We have seen 4 - byte - aligned buffers allocated on
+            // virtual x64/x86 machines which were accepted by HttpReceiveHttpRequest without errors. In
+            // these cases the buffer alignment may cause reading values at invalid offset. Setting buffer
+            // alignment to 0 for now.
+            // 
+            // _bufferAlignment = (int)(requestAddress.ToInt64() & 0x07);
+            _bufferAlignment = 0;
+
+            var newSize = (int)(bufferSize ?? DefaultBufferSize) + AlignmentPadding;
+            if (newSize <= memoryPool.MaxBufferSize)
+            {
+                _backingBuffer = memoryPool.Rent(newSize);
+            }
+            else
+            {
+                // No size limit
+                _backingBuffer = MemoryPool<byte>.Shared.Rent(newSize);
```

Comment:
IIRC ArrayPool<byte>.Shared is better than MemoryPool.Shared.Rent because it can be devirtualized. Also, according to @davidfowl. IMemoryPool allocates a lot.

I know it would be awkward as ArrayPool.Rent returns byte[] rather than IMemoryOwner, you would need to do a check like this: https://github.com/dotnet/corefx/blob/master/src/System.IO.Pipelines/src/System/IO/Pipelines/BufferSegment.cs#L64

I'm not necessarily suggesting we do this, but food for thought.

Reply:
Yes, it avoids the allocation.

Reply:
MemoryPool<byte>.Shared is just a wrapper for ArrayPool.Shared. This code path does not warrant that extra complexity at this time.

Reply:
Note this only gets hit as a fallback for large requests. Mosts requests go through the slab pool.